README.md
# elevator_project
4층 엘리베이터 모형 졸작 프로젝트 파일 저장소

※ 프로그램 개발을 마무름.

개발 시작 ~ 최종본까지의 진행 기록은 change_log 파일을 참조


change_log
2021년 4월 25일
핀 번호 정의 거의 완료. 로직 구현은 되어 있지 않음

2021년 5월 1일
- 알고리즘 반영 시작
- 핀번호 전처리 변경사항 있음
- 전역변수 수정/추가
- Stepper 라이브러리 BYJ 모터 선언 추가
- setup 함수 작성 완료(검토 필요)
- ISR 함수 중 호출/목적지 버튼 부분 작성 완료(검토 필요)
- irGetDist 함수 기능 및 명칭 변경 -> getFloor, 층수 구하고 FND로 표시
- 74HC595 쉬프트 레지스터로 FND 구동, shiftOut 함수 사용 결정(확정사항 아님)
- loop 함수에 알고리즘 구현 시작
- 오류 발견: UP과 DN 전처리 부분이 각각 두 번 선언되어 있음. 수정 요

2021년 5월 6일
- 일부 중요한 함수 구현 시작(모터 제어, 이동방향 결정 함수 등)
- BYJ에서 NEMA로 바꿀 수 있다는 것을 감안해 주석화한 코드를 추가함.
- setup 함수 등 일부 함수에 변경사항이 있음.
- 자잘한 오류 수정

2021년 7월 28일
• IR 센서를 모두 디지털로 처리하도록 변경함
• 문짝 구동 모터를 byj에서 네마 모터 처리로 코드를 변경함
• 목적지 확인 함수(chkDest) 일부 작성
• 모터 정밀제어 함수 일부 작성 (디버깅 필요)
• loop 함수 수정
• 전역 변수 수정
• getFloor 함수 수정(층수 표시 구현 안됨)
• 메인 모터 구동 함수에 운전 상태(상향? 하향?) 기록 부분 추가
• 문 구동 함수 작성(미완, 모터 방향을 확인할 필요가 있음)

2021년 7월 30일
• FND 제어 함수를 getFloor 함수에서 사용하도록 함(층수 표시가 가능해짐)
• getFloor 함수가 loop 함수에서 목적지/호출지가 아니면서 카 모터가 구동중인 경우 작동되도록 함
• chkDest 함수에서 목적지/호출지 배열을 조작하도록 하고, 반환값을 지정함
• loop 함수 수정함(carStat 조정 부분 삭제, 로직 수정 등)
• motorDrv 함수에서 isMoving(움직이는가?) 변수를 조작하도록 함.
• 주석문을 수정함
※ 비상 버튼 기능을 아직 구현하지 않았음. 비상 버튼은 인터럽트가 아닌 loop 함수 내에서 구현할 예정.

2021년 8월 3일
• 비상운전을 구현함(디버깅 필요)
• 비상운전 구현을 위해 loop함수 내 제어 코드를 별도의 함수로 분리함.
• 문 제어 함수를 호출하는 코드에 문제가 있어 이를 바로잡음.
• setup 함수에서 잘못된 부분이 있어 이를 바로잡음.
※ 엘리베이터 실물의 모터 구동 방향을 아직 확인하지 않았음. 이는 추후 반영할 것임.

2021년 10월 26일
■ 추가
• 전처리 부분 카 움직임 명령값/상태값 추가(상승-저속복귀운전, 하강-저속복귀운전)
• 74HC595 IC(FND 제어) 핀 번호 정의 추가
• spdStat 전역변수 추가
→ 비상정지 기능 구현을 위해 가속·감속·정속·저속복귀 상태를 기록하는 변수
• motorDrv 함수에 비상정지 후 복귀 운전을 추가

■ 수정
• 주석을 조금 더 알아보기 좋게 수정
• setup 함수 수정(핀 번호 전처리 바뀐 부분을 반영함)
• doorDrv 함수와 setup에서 적외선 센서로 감지하는 대신 일정 스텝각 만큼 돌아가게 바꿈(하드웨어 구성을 반영함)
☆ 소방운전을 비상정지로 바꿈
→ 미완성인 기능 대신 유지보수/디버그 작업에 쓰기 좋은 기능으로 갈음함

■ 삭제
• chkDoor 함수 삭제(문 상태 센서가 없으므로 필요가 없음)
• 사용하지 않는 핀 번호 정의(LEDBn, IR_SNSR_DOOR_ ……)를 삭제하거나 주석처리(하드웨어 구성 반영)
• irStat, isEmergency, isDoorOpen 전역변수 삭제(필요 없음)
• 없어진 핀 번호 정의가 들어간 코드를 모두 지움

■ 해야할 것
※ 비상상황 대처를 위해 시험운전은 모두 랩 파워서플라이 출력을 11.9V로 설정하고 진행하며, 메인 파워서플라이를 쓰지 않는다.
※ 비상상황이 발생하면 바로 랩 파워서플라이의 전압값을 내리거나 전류값을 내려 작동을 멈추고, 카가 천천히 내려오도록 손으로 잡는다.
• 바뀐 doorDrv 함수가 작동하는가?
• 대체적으로 잘 작동하는가?

motor_tst2.c
#include <HCMotor.h>
#include <Stepper.h>

#define NEMA_DIR 8
#define NEMA_CLK 9 
#define BYJ1 34
#define BYJ2 35
#define BYJ3 36
#define BYJ4 37

#define STEPS 2047

const int byjSteps = 2047;
const int spd = 25, nemaMainSpd = 25, nemaMainSlowSpd = 70;
const int nemaAccDelay = 25, nemaDeccDelay = 25;

HCMotor nema;
Stepper byj(STEPS, BYJ1, BYJ3, BYJ2, BYJ4);

void setup() {
  // put your setup code here, to run once:
  nema.Init();
  nema.attach(0, STEPPER, NEMA_CLK, NEMA_DIR);
  nema.Steps(0, CONTINUOUS);
  nema.DutyCycle(0, spd); /* 10 to 1024 */
  byj.setSpeed(15);
}

void loop() {
  // put your main code here, to run repeatedly:
  int spdTmpAcc = nemaMainSlowSpd;
  nema.Direction(0, REVERSE);
  while (spdTmpAcc > nemaMainSpd) {
      nema.DutyCycle(0, spdTmpAcc--);
      delay(nemaAccDelay);
  }        
  nema.DutyCycle(0, spd);  
  nema.Direction(0, REVERSE);
  delay(2000);
  int spdTmpDecc = nemaMainSpd;
  nema.Direction(0, REVERSE);
  while (spdTmpDecc < nemaMainSlowSpd) {
      nema.DutyCycle(0, spdTmpDecc++);
      delay(nemaDeccDelay);
  }
  delay(1000);      
  nema.DutyCycle(0, 0);
  delay(1000);
  spdTmpAcc = nemaMainSlowSpd;
  nema.Direction(0, FORWARD);
  while (spdTmpAcc > nemaMainSpd) {
      nema.DutyCycle(0, spdTmpAcc--);
      delay(nemaAccDelay);
  }        
  nema.DutyCycle(0, spd);
  nema.Direction(0, FORWARD);
  delay(2000);
  spdTmpDecc = nemaMainSpd;
  nema.Direction(0, FORWARD);
  while (spdTmpDecc < nemaMainSlowSpd) {
      nema.DutyCycle(0, spdTmpDecc++);
      delay(nemaDeccDelay);
  }
  delay(1000);
  nema.DutyCycle(0, 0);
  delay(1000);
  byj.step(-byjSteps);
  delay(1000);
  byj.step(byjSteps);
}


debug_code.c
/* 디버그용 파일! 시리얼 통신으로 터미널에 함수 실행 상태를 실시간으로 보냄 */

/* Atmega2560 마이크로 컨트롤러는 8비트이므로 기본 데이터 타입은 최적화를 위해 바이트(unsigned char)로 함. */

/* 
네마모터는 시계방향 회전이 역방향 회전인 것으로 보임. 결선에 따라 달라질 수 있으나, 카 모터 시험 코드의 방향은 그렇게 되어 있었음.
네마모터는 1 스텝이 1.8 ± 5% 이므로, 200스텝 정도 돌리면 한 바퀴 돌아감. 문을 개폐할 때 센서를 사용할 수 없는 경우 스텝각 150~250 사이의 적정값을 찾도록 함.
*/

/*
아두이노 메가 핀 번호
인터럽트 핀(순서대로): 2, 3, 21, 20, 19, 18
PWM 핀: 2~13, 44~46
디지털 핀: 0 ~ 53
아날로그 입력: A0~A15
I2C 통신 핀: 20(SCA), 21(SCL)
직렬통신 핀: 0, 1, 14~19
*/

/*
• 버튼 아날로그 검출용으로 저항을 쓸 경우 저항값은 10kOhm으로 한다.
• 호출 버튼과 목적지 버튼은 인터럽트 구현을 위해 모두 인터럽트 핀에 연결한다. 다이오드를 사용한다(오류 방지)
*/

/* 코드 시작 부분 */

/* 헤더 불러오기 */
#include <HCMotor.h>
/* 라이브러리 설명: https://forum.hobbycomponents.com/viewtopic.php?f=58&t=1870 */
/* TB6560 제어 http://scipia.co.kr/cms/blog/225 */

/* 참과 거짓 전처리 */
#define FALSE 0
#define TRUE 1

/* 명령어 정의(전처리) */
#define STOP 0
#define UP 1
#define UP_SLOW 2
#define UP_ACCEL 3
#define DN 4
#define DN_SLOW 5
#define DN_ACCEL 6
#define UP_RECY 8
#define DN_RECY 9
#define CLOSE 0
#define OPEN 1

/* 디지털 핀 번호 정의 */
#define BTN_CALL_INTR 2 /* 호출 버튼 인터럽트 */
#define BTN_DEST_INTR 3 /* 목적지 버튼 인터럽트 */

#define NEMA_DOOR_DIR 5 /* CW+ */
#define NEMA_DOOR_STEP 6 /* CLK+ */

/* TB6560 드라이버의 CW-, CLK- 단자는 모두 아두이노의 GND에 연결한다. */
#define NEMA_MAIN_DIR 8 /* CW+ */
#define NEMA_MAIN_STEP 9 /* CLK+ */

#define HC595_DATA 10 /* FND 구동용 쉬프트 레지스터 */
#define HC595_LATCH 11
#define HC595_CLOCK 12

#define LEDF1U 22 /* 각 층의 호출버튼 LED 핀번호 정의 */
#define LEDF2U 23
#define LEDF2D 24
#define LEDF3U 25
#define LEDF3D 26
#define LEDF4D 27
#define LEDUP 32 /* 운전 방향 표시등 정의 */
#define LEDDN 33

#define IR_SNSR_1 34 /* 카 위치 감지 거리 센서 정의 */
#define IR_SNSR_2 35
#define IR_SNSR_3 36
#define IR_SNSR_4 37
#define BTN1U 38 /* 층별 호출 버튼 정의 */
#define BTN2U 39
#define BTN2D 40
#define BTN3U 41
#define BTN3D 42
#define BTN4D 43
#define BTN1 44 /* 카 목적지 버튼 정의 */
#define BTN2 45
#define BTN3 46
#define BTN4 47
/*
#define IR_SNSR_DOOR_OPEN 48
#define IR_SNSR_DOOR_CLOSE 49
*/
#define BTN_EMERGENCY 50 /* 비상버튼 */
#define LIGHT_EMERGENCY 51 /* 비상운전 표시등 */

/* pinouts (analog) */

/*
#define BTN_CALL "A0"
#define BTN_DEST "A1"
*/

/* 디지털 구현이 어려운 경우 적외선 센서는 아날로그로 구현할 수 있도록 한다.  */

typedef unsigned char BYTE; /* 바이트는 부호 없는 char */
typedef BYTE BOOL; /* 부울 자료형의 형식을 바이트로 지정 */

volatile BOOL arrDest[] = { FALSE, FALSE, FALSE, FALSE }; /* 목적지 입력 상태 저장용 배열 */
volatile BOOL arrCall[] = { FALSE, FALSE, FALSE, FALSE, FALSE, FALSE }; 
/* 호출 버튼 상태 저장용 배열, F1U, F2U, F3U, F2D, F3D, F4D 순 */
volatile BYTE spdStat = STOP; /* 정속/감속/가속/저속복귀 상태 기록 */
volatile BOOL isMoving = FALSE; /* 움직이는가? */
volatile BYTE carStat = STOP; /* 카 운전 상태(방향) */
volatile BYTE currFloor = 1; /* 현재 위치 */
volatile BOOL doorStat = CLOSE; /* 문 상태 변수(필요 없을 시 제거) */

const BYTE fndDigits[] = { 0x03, 0x9F, 0x25, 0x0D, 0x99 }; /* 0~4 */
const int nemaMainSpd = 40; /* 10~1024, 낮을 수록 빠름 */
const int nemaMainSlowSpd = 500;
const int nemaDoorSpd = 500;
/* const int nemaDoorSteps = 200  */

HCMotor nemaMain;
HCMotor nemaDoor;

void getFloor(void);
void motorDrv(BYTE drvMode);
void doorDrv(BOOL op);
BOOL chkDest(void);
BYTE chkUpDn(void);
void preciseMotorCtrl(BYTE floor);
void fndDrv(BYTE floorNum);
void ctrlMove(BYTE mode);
void strtr(void);

/* 인터럽트 함수 */
void isrBtnCall(void) { /* 호출 버튼 인터럽트 처리 */
    Serial.println("[isrBtnCall] Execution");
    /* int analogVal = analogRead(BTN_CALL); */
    if (digitalRead(BTN1U)) {
        Serial.println("[isrBtnCall] 1U INPUT");
        arrCall[0] = TRUE;
        digitalWrite(LEDF1U, HIGH);
    }
    if (digitalRead(BTN2U)) {
        Serial.println("[isrBtnCall] 2U INPUT");
        arrCall[1] = TRUE;
        digitalWrite(LEDF2U, HIGH);
    }
    if (digitalRead(BTN2D)) {
        Serial.println("[isrBtnCall] 2D INPUT");
        arrCall[2] = TRUE;
        digitalWrite(LEDF2D, HIGH);
    }
    if (digitalRead(BTN3U)) {
        Serial.println("[isrBtnCall] 3U INPUT");
        arrCall[3] = TRUE;
        digitalWrite(LEDF3U, HIGH);
    }
    if (digitalRead(BTN3D)) {
        Serial.println("[isrBtnCall] 3D INPUT");
        arrCall[4] = TRUE;
        digitalWrite(LEDF3D, HIGH);
    }
    if (digitalRead(BTN4D)) {
        Serial.println("[isrBtnCall] 4D INPUT");
        arrCall[5] = TRUE;
        digitalWrite(LEDF4D, HIGH);
    }
}

void isrBtnDest(void) { /* 목적층 버튼 인터럽트 처리 */
    Serial.println("[isrBtnDest] Execution");
    /* int analogVal = analogRead(BTN_DEST); */
    if (digitalRead(BTN1)) {
        Serial.println("[isrBtnDest] 1 INPUT");
        arrDest[0] = TRUE;
    }
    if (digitalRead(BTN2)) {
        Serial.println("[isrBtnDest] 2 INPUT");
        arrDest[1] = TRUE;
    }
    if (digitalRead(BTN3)) {
        Serial.println("[isrBtnDest] 3 INPUT");
        arrDest[2] = TRUE;
    }
    if (digitalRead(BTN4)) {
        Serial.println("[isrBtnDest] 4 INPUT");
        arrDest[3] = TRUE;
    }
}

void setup() {
    Serial.begin(9600);
    Serial.println("Serial Comm initialized");
    /* 핀 배열에 맞게 pinout 정의 */
    /* 인터럽트 모두 실행 */
    pinMode(BTN_CALL_INTR, INPUT);
    pinMode(BTN_DEST_INTR, INPUT);
    pinMode(HC595_DATA, OUTPUT);
    pinMode(HC595_CLOCK, OUTPUT);
    pinMode(HC595_LATCH, OUTPUT);
    pinMode(LEDF1U, OUTPUT);
    pinMode(LEDF2U, OUTPUT);
    pinMode(LEDF2D, OUTPUT);
    pinMode(LEDF3U, OUTPUT);
    pinMode(LEDF3D, OUTPUT);
    pinMode(LEDF4D, OUTPUT);
    pinMode(LEDUP, OUTPUT);
    pinMode(LEDDN, OUTPUT);
    pinMode(BTN1U, INPUT);
    pinMode(BTN2U, INPUT);
    pinMode(BTN2D, INPUT);
    pinMode(BTN3U, INPUT);
    pinMode(BTN3D, INPUT);
    pinMode(BTN4D, INPUT);
    pinMode(BTN1, INPUT);
    pinMode(BTN2, INPUT);
    pinMode(BTN3, INPUT);
    pinMode(BTN4, INPUT);
    pinMode(IR_SNSR_1, INPUT);
    pinMode(IR_SNSR_2, INPUT);
    pinMode(IR_SNSR_3, INPUT);
    pinMode(IR_SNSR_4, INPUT);
    pinMode(BTN_EMERGENCY, INPUT);
    pinMode(LIGHT_EMERGENCY, OUTPUT);
    attachInterrupt(0, isrBtnCall, RISING);
    attachInterrupt(1, isrBtnDest, RISING);
    nemaMain.Init();
    nemaMain.attach(0, STEPPER, NEMA_MAIN_STEP, NEMA_MAIN_DIR);
    nemaMain.Steps(0, CONTINUOUS);
    nemaMain.DutyCycle(0, 0);
    nemaDoor.Init();
    nemaDoor.attach(1, STEPPER, NEMA_DOOR_STEP, NEMA_DOOR_DIR);
    nemaDoor.Steps(1, 190); /* 문 모터는 한 바퀴보다 살짝 덜 돌아가게 설정 */
    nemaDoor.DutyCycle(1, 0);
    getFloor();
    Serial.println("Setup complete");
}

void loop(void) { 
    if (digitalRead(BTN_EMERGENCY)) { /* 비상정지 버튼 입력 */
        Serial.println("[loop] Emergency btn input");
        nemaMain.DutyCycle(0, 0);
        digitalWrite(LIGHT_EMERGENCY, HIGH);
        Serial.println("[loop] In infinite loop for Emergency stop");
        while(!digitalRead(BTN_EMERGENCY)) /* 버튼 릴리스 전까지 무한 빈 루프 */
            (void) "aoeu";
        digitalWrite(LIGHT_EMERGENCY, LOW); /* 버튼 릴리스 뒤 소등 및 모터 재구동 */
        Serial.println("[loop] Infinite loop(emergency stop) ended");
        if (carStat == UP && spdStat == UP) motorDrv(UP);
        if (carStat == UP && spdStat == UP_ACCEL) motorDrv(UP);
        if (carStat == UP && spdStat == UP_SLOW) motorDrv(UP_RECY);
        if (carStat == UP && spdStat == UP_RECY) motorDrv(UP_RECY);
        if (carStat == DN && spdStat == DN) motorDrv(DN);
        if (carStat == DN && spdStat == DN_ACCEL) motorDrv(DN);
        if (carStat == DN && spdStat == DN_SLOW) motorDrv(DN_RECY);
        if (carStat == DN && spdStat == DN_RECY) motorDrv(DN_RECY);
        return;
    }
    if (isMoving) {
        Serial.println("[loop] isMoving var is TRUE");
        ctrlMove(1);
        return;
    } else {
        Serial.println("[loop] isMoving var is FALSE");
        strtr();
        return;
    }
}

void getFloor(void) { /* 층수 구하고 FND로 표시하는 함수 */
    Serial.println("[getFloor] Execution");
    if (digitalRead(IR_SNSR_1)) {
        Serial.println("[getFloor] Floor is 1");
        currFloor = 1;
        fndDrv(1);
    }
    else if (digitalRead(IR_SNSR_2)) {
        Serial.println("[getFloor] Floor is 2");
        currFloor = 2;
        fndDrv(2);
    }
    else if (digitalRead(IR_SNSR_3)) {
        Serial.println("[getFloor] Floor is 3");
        currFloor = 3;
        fndDrv(3);
    }
    else if (digitalRead(IR_SNSR_4)) {
        Serial.println("[getFloor] Floor is 4");
        currFloor = 4;
        fndDrv(4);
    }
    /* 센서 고장 시의 운전은 구현되지 않았음 */
}

void motorDrv(BYTE drvMode) { /* 모터 구동 */
    Serial.println("[motorDrv] Execution");
    int spdTmp = 0;
    switch(drvMode) {
        case STOP:
        Serial.println("[motorDrv] drvMode is STOP");
        nemaMain.DutyCycle(0, 0);
        carStat = STOP;
        spdStat = STOP;
        isMoving = FALSE;
        break;

        case UP:
        Serial.println("[motorDrv] drvMode is UP");
        nemaMain.Direction(0, REVERSE);
        nemaMain.DutyCycle(0, nemaMainSpd);
        carStat = UP;
        spdStat = UP;
        isMoving = TRUE;
        break;

        case DN:
        Serial.println("[motorDrv] drvMode is DN");
        nemaMain.Direction(0, FORWARD);
        nemaMain.DutyCycle(0, nemaMainSpd);
        carStat = DN;
        spdStat = DN;
        isMoving = TRUE;
        break;

        /* 목적층/호출위치에 가까워졌을 때 속도를 줄이고 정밀하게 운전 */
        case UP_SLOW:
        Serial.println("[motorDrv] drvMode is UP_SLOW");
        spdTmp = nemaMainSpd;
        nemaMain.Direction(0, FORWARD);
        while (spdTmp == nemaMainSlowSpd) {
            nemaMain.DutyCycle(0, spdTmp++);
            delay(1);
        }
        spdStat = UP_SLOW;
        break;

        case UP_ACCEL:
        Serial.println("[motorDrv] drvMode is UP_ACCEL");
        spdTmp = 1000;
        nemaMain.Direction(0, FORWARD);
        while (spdTmp == nemaMainSpd) {
            nemaMain.DutyCycle(0, spdTmp--);
            delay(1);
        }        
        spdStat = UP_ACCEL;
        break;

        case DN_SLOW:
        Serial.println("[motorDrv] drvMode is DN_SLOW");
        spdTmp = nemaMainSpd;
        nemaMain.Direction(0, REVERSE);
        while (spdTmp == nemaMainSlowSpd) {
            nemaMain.DutyCycle(0, spdTmp++);
            delay(1);
        }        
        spdStat = DN_SLOW;
        break;

        case DN_ACCEL:
        Serial.println("[motorDrv] drvMode is DN_ACCEL");
        spdTmp = 1000;
        nemaMain.Direction(0, REVERSE);
        while (spdTmp == nemaMainSpd) {
            nemaMain.DutyCycle(0, spdTmp--);
            delay(1);
        }        
        spdStat = DN_ACCEL;
        break;

        /* 저속 복귀 운전 명령 */
        case UP_RECY:
        Serial.println("[motorDrv] drvMode is UP_RECY");
        nemaMain.Direction(0, FORWARD);
        nemaMain.DutyCycle(0, nemaMainSlowSpd);
        spdStat = UP_RECY;
        break;

        case DN_RECY:
        Serial.println("[motorDrv] drvMode is DN_RECY");
        nemaMain.Direction(0, REVERSE);
        nemaMain.DutyCycle(0, nemaMainSlowSpd);
        spdStat = DN_RECY;
        break;
    }
}

void doorDrv(BOOL op) { /* 문 구동용 함수 */
    Serial.println("[doorDrv] Execution");
    if (doorStat == OPEN && op == CLOSE) {
        Serial.println("[doorDrv] Closing");
        /* 문을 닫는다  */
        nemaDoor.Direction(1, REVERSE);
        nemaDoor.DutyCycle(1, nemaDoorSpd);
        doorStat = CLOSE;
        nemaDoor.DutyCycle(1, 0);
        return;
    }
    else if (doorStat == CLOSE && op == OPEN) {
        Serial.println("[doorDrv] Opening");
        nemaDoor.Direction(1, FORWARD);
        nemaDoor.DutyCycle(1, nemaDoorSpd);
        doorStat = OPEN;
        nemaDoor.DutyCycle(1, 0);
        return;
    }
    Serial.println("[doorDrv] ?Problem occured(doorStat does not match with op)")
    else if (doorStat == CLOSE && op == CLOSE) return;
    else if (doorStat == OPEN && op == OPEN) return;
}

BOOL chkDest(void) { /* 현재 층수가 호출된/목적지 층수인지를 확인한다. */
    Serial.println("[chkDest] Execution");
    if (digitalRead(IR_SNSR_1) == 1 && (arrDest[0] == TRUE || arrCall[0] == TRUE)) {
        Serial.println("[chkDest] SNSR 1 INPUT");
        preciseMotorCtrl(1);
        arrCall[0] = FALSE;
        arrDest[0] = FALSE;
        return TRUE;
    }
    else if (digitalRead(IR_SNSR_2) == 1 && (arrDest[1] == TRUE || arrCall[1] == TRUE && carStat == UP || arrCall[3] && carStat == DN)) {
        Serial.println("[chkDest] SNSR 2 INPUT");
        preciseMotorCtrl(2);
        if (carStat == UP) arrCall[1] = FALSE;
        else if (carStat == DN) arrCall[3] = FALSE;
        arrDest[1] = FALSE;
        return TRUE;
    }
    else if (digitalRead(IR_SNSR_3) == 1 && (arrDest[2] == TRUE || arrCall[2] && carStat == UP || arrCall[4] && carStat == DN)) {
        Serial.println("[chkDest] SNSR 3 INPUT");
        preciseMotorCtrl(3);
        if (carStat == UP) arrCall[2] = FALSE;
        else if (carStat == DN) arrCall[4] = FALSE;
        arrDest[2] = FALSE;
        return TRUE;
    }
    else if (digitalRead(IR_SNSR_4) == 1 && (arrDest[3] == TRUE || arrCall[3] == TRUE)) {
        Serial.println("[chkDest] SNSR 4 INPUT");
        preciseMotorCtrl(4);
        arrCall[5] = FALSE;
        arrDest[3] = FALSE;
        return TRUE;
    }
    Serial.println("[chkDest] NO SNSR INPUT");
    return FALSE;
}

BYTE chkUpDn(void) { /* 상승/하강 결정 함수 */ /* 디버깅 필요 */
    Serial.println("[chkUpDn] Execution");
    /* 상승? */
    if (carStat != DN) {
        for (int i = (currFloor - 1); i < 4; i++) {
            if (arrDest[i] && i > currFloor) return UP; 
        }
        for (int i = (currFloor - 1); i < 4; i++) {
            if (arrCall[i] && i > currFloor) return UP;
        }
    }
    /* 하강? */
    if (carStat != UP) {
        for (int i = (currFloor - 1); i >= 0; i--) {
            if (arrDest[i] && i < currFloor) return DN;
        }
        for (int i = (currFloor - 1); i > 2; i--) {
            if (arrCall[i] && i < currFloor) return DN;
        }
    }
    /* 이동할 필요 없음 */
    return STOP;
}

void preciseMotorCtrl(BYTE floor) { /* 모터 정밀제어 함수 */
    Serial.println("[preciseMotorCtrl] Execution");
    if (carStat == UP) motorDrv(UP_SLOW);
    else if (carStat == DN) motorDrv(DN_SLOW);

    Serial.println("[preciseMotorCtrl] Motor slowed down");
    switch(floor) { /* 센서 입력이 들어올 때까지 무한 빈 루프 */
        case 1:
        while (digitalRead(IR_SNSR_1)) (void) "aoeu";
        break;

        case 2:
        while (digitalRead(IR_SNSR_2)) (void) "aoeu";
        break;

        case 3:
        while (digitalRead(IR_SNSR_3)) (void) "aoeu";
        break;

        case 4:
        while (digitalRead(IR_SNSR_4)) (void) "aoeu";
        break;
    }
    Serial.println("[preciseMotorCtrl] Stopping motor");
    motorDrv(STOP);
    return;
}

void fndDrv(BYTE floorNum) { /* 7seg 구동 */
    Serial.println("[fndDrv] Execution");
    digitalWrite(HC595_LATCH, LOW);
    shiftOut(HC595_DATA, HC595_CLOCK, LSBFIRST, fndDigits[floorNum]);
    digitalWrite(HC595_LATCH, HIGH);
} 

void ctrlMove(BYTE mode) { /* mode 0은 정지(버튼만 눌림), 1은 움직이는 중에 제어 */
    Serial.println("[ctrlMove] Execution");
    if (!mode) goto drivedoor;
    chkDest();
    if (isMoving) {
        Serial.println("[ctrlMove] isMoving is TRUE");
        getFloor();
        return;
	}
    Serial.println("[ctrlMove] isMoving is FALSE");
    switch (currFloor) {
        case 1:
        digitalWrite(LEDF1U, LOW);
        break;

        case 2:
        if (carStat == UP) digitalWrite(LEDF2U, LOW);
        else if (carStat == DN) digitalWrite(LEDF2D, LOW);
        break;

        case 3:
        if (carStat == UP) digitalWrite(LEDF3U, LOW);
        else if (carStat == DN) digitalWrite(LEDF3D, LOW);
        break;

        case 4:
        digitalWrite(LEDF4D, LOW);
        break;
    }
    drivedoor:
    Serial.println("[ctrlMove] Driving door...");
    if (doorStat == CLOSE) { 
        Serial.println("[ctrlMove] Opening door...");
        doorDrv(OPEN);
        delay(5000);
    }
    doorDrv(CLOSE);
    Serial.println("[ctrlMove] Closed door");
    return;
}

void strtr(void) {
    Serial.println("[strtr] Execution");
    for (int i = 0; i < 4; i++) { /* 움직여야 하는가?, 목적지 입력 확인 */
        if (arrDest[i]) goto start;
    }
    for (int i = 0; i < 6; i++) { /* 움직여야 하는가?, 호출 입력 확인 */
        if (arrCall[i]) goto start;
    }
    /* 대기한다 */
    Serial.println("[strtr] Staying");
    digitalWrite(LEDUP, LOW);
    digitalWrite(LEDDN, LOW);
    return;
    
    /* 시동한다  */
    start:
    Serial.println("[strtr] Starting");
    BYTE updn = chkUpDn();
    if (updn == UP) {
        Serial.println("[strtr] UP");
        digitalWrite(LEDUP, HIGH);
        digitalWrite(LEDDN, LOW);
        motorDrv(UP_ACCEL);
        motorDrv(UP);
    } else if (updn == DN) {
        Serial.println("[strtr] DN");
        digitalWrite(LEDUP, LOW);
        digitalWrite(LEDDN, HIGH);
        motorDrv(DN_ACCEL);
        motorDrv(DN);
    } else if (updn == STOP) {
        Serial.println("[strtr] STOP");
        ctrlMove(0);
    }
}
